oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/JdbcOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,514 @@
+ * @see JdbcOperations
+ * @see RowMapper
+ */
+public class JdbcOAuth2AuthorizedClientService implements OAuth2AuthorizedClientService {
+	private static final String COLUMN_NAMES =
+			"client_registration_id, " +
+			"principal_name, " +
+			"access_token_type, " +
+			"access_token_value, " +
+			"access_token_issued_at, " +
+			"access_token_expires_at, " +
+			"access_token_scopes, " +
+			"refresh_token_value, " +
+			"refresh_token_issued_at";
+	private static final String TABLE_NAME = "oauth2_authorized_client";
+	private static final String PK_FILTER = "client_registration_id = ? AND principal_name = ?";
+	private static final String LOAD_AUTHORIZED_CLIENT_SQL = "SELECT " + COLUMN_NAMES +
+			" FROM " + TABLE_NAME + " WHERE " + PK_FILTER;
+	private static final String SAVE_AUTHORIZED_CLIENT_SQL = "INSERT INTO " + TABLE_NAME +
+			" (" + COLUMN_NAMES + ") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)";
+	private static final String REMOVE_AUTHORIZED_CLIENT_SQL = "DELETE FROM " + TABLE_NAME +
+			" WHERE " + PK_FILTER;
+	protected final JdbcOperations jdbcOperations;
+	protected RowMapper<OAuth2AuthorizedClientData> authorizedClientDataRowMapper;
+	protected Converter<OAuth2AuthorizedClientData, OAuth2AuthorizedClient> authorizedClientDataConverter;
+	protected Converter<OAuth2AuthorizedClientHolder, OAuth2AuthorizedClientData> authorizedClientConverter;
+
+	/**
+	 * Constructs a {@code JdbcOAuth2AuthorizedClientService} using the provided parameters.
+	 *
+	 * @param jdbcOperations the JDBC operations
+	 * @param clientRegistrationRepository the repository of client registrations
+	 */
+	public JdbcOAuth2AuthorizedClientService(
+			JdbcOperations jdbcOperations, ClientRegistrationRepository clientRegistrationRepository) {
+
+		Assert.notNull(jdbcOperations, "jdbcOperations cannot be null");
+		Assert.notNull(clientRegistrationRepository, "clientRegistrationRepository cannot be null");
+		this.jdbcOperations = jdbcOperations;
+		this.authorizedClientDataRowMapper = new OAuth2AuthorizedClientDataRowMapper();
+		this.authorizedClientDataConverter = new OAuth2AuthorizedClientDataConverter(clientRegistrationRepository);
+		this.authorizedClientConverter = new OAuth2AuthorizedClientConverter();
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T extends OAuth2AuthorizedClient> T loadAuthorizedClient(String clientRegistrationId, String principalName) {
+		Assert.hasText(clientRegistrationId, "clientRegistrationId cannot be empty");
+		Assert.hasText(principalName, "principalName cannot be empty");
+
+		Object[] args = {clientRegistrationId, principalName};
+		int[] argTypes = {Types.VARCHAR, Types.VARCHAR};
+		PreparedStatementSetter pss = new ArgumentTypePreparedStatementSetter(args, argTypes);
+
+		List<OAuth2AuthorizedClientData> result = this.jdbcOperations.query(
+				LOAD_AUTHORIZED_CLIENT_SQL, pss, this.authorizedClientDataRowMapper);
+
+		return !result.isEmpty() ? (T) this.authorizedClientDataConverter.convert(result.get(0)) : null;
```

Comment:
Since we are only looking for 0 or 1 results, it seems like queryForObject might make more sense here.

Reply:
I used `queryForObject()` originally but it throws an `IncorrectResultSizeDataAccessException` if not found, which is not ideal. With `List query()`, I can avoid the `try catch` and exception being thrown.